### PR TITLE
Upload images via a pre-signed URL

### LINKED ISSUE #308: Allow image uploads with pre-signed url
The image uploads need to work with a pre-signed URL
To make it happen, the editor can expect direct uploads URL.

```js
<Editor 
  directUploadsURL={`api/v1/direct_uploads`} 
/>
```

This direct upload URL will give a pre-signed URL to which the editor can upload the actual file.
To see how direct uploads work, refer following:

DirectUploads definition in Rails
https://github.com/rails/rails/blob/aaa64687e81873b26c964fada0b969621ae8cf8f/activestorage/app/javascript/activestorage/direct_upload.js#L7

How that's used: 

https://github.com/bigbinary/neeto-planner-web/blob/1d188604cf5d77f43dbb91f3cf38cb9f129d67e1/app/javascript/src/components/Layouts/Header/Attachment.jsx#L36-L46

### PR DESCRIPTION
**Pull request by AbhayVAshokan**
Fixes #308 

**Description**

- Added: **BREAKING** support for pre-signed URLs for uploading images.

**Checklist**

- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] I have added the necessary label (patch/minor/major - If package publish
      is required).
- [x] I have followed the suggested description format and styling.

**Reviewers**

@AbhayVAshokan _a

### CODE REVIEW

File: lib/components/Editor/CustomExtensions/Image/LocalUploader.jsx [hunk 1]
```diff
@@ -46,9 +40,15 @@ const LocalUploader = ({
...
-      .use(XHRUpload, { endpoint, ...UPPY_UPLOAD_CONFIG })
+      .use(ActiveStorageUpload, {
+        directUploadUrl: DIRECT_UPLOAD_ENDPOINT,
+        ...UPPY_UPLOAD_CONFIG,
+      })
       .on("upload", () => setIsUploading(true))
-      .on("upload-success", (_, response) => onSuccess(response.body.imageURL))
+      .on("upload-success", ({ data: { name } }, { signed_id }) => {
+        const imageUrl = `${window.location.origin}/rails/active_storage/blobs/redirect/${signed_id}/${name}`;
+        onSuccess(imageUrl);
```
Comment (AbhayVAshokan) [inline lib/components/Editor/CustomExtensions/Image/LocalUploader.jsx:50]: @sandip-mane _a I need your expertise in this issue. Could you please take a look at it and provide your thoughts?
https://vimeo.com/744279451/12227d0cac

Comment (sandip-mane) [reply lib/components/Editor/CustomExtensions/Image/LocalUploader.jsx:48]: @AbhayVAshokan https://app.usebubbles.com/ah2eZDQftYg8LxUAgo6Tzo/file-path

Comment (AbhayVAshokan) [reply lib/components/Editor/CustomExtensions/Image/LocalUploader.jsx:48]: Oh yes. Great idea. Let me make the required changes.

**Comment by AbhayVAshokan** [comment]
@labeebklatif _a please review.

https://vimeo.com/744615877/9edc7df2ed


References:
1. [uppy-active-storage-upload](https://github.com/excid3/uppy-activestorage-upload)
2. [AWSS3Multipart](https://github.com/transloadit/uppy/blob/main/packages/@uppy/aws-s3-multipart/src/index.js)
3. [direct_uploads_controller.rb](https://github.com/rails/rails/blob/main/activestorage/app/controllers/active_storage/direct_uploads_controller.rb)

cc: @karthiknmenon @sandip-mane

**Comment by labeebklatif** [comment]
@karthiknmenon With `@rails/activestorage` added as a peer dependency, the editor will be restricted to only rails based projects, rights? Is this intended?

**Comment by AbhayVAshokan** [comment]
@labeebklatif 

In that case, `@rails/activestorage` can be removed from peer dependencies and it can be bundled along with the editor.

My point was: that if we had a custom implementation for the `@rails/activestorage` package or not, we would still be performing the same set of operations such as generating the checksum, calling the first API for the pre-signed URL, and calling a PUT request to S3 bucket.

I had looked into a few implementations in NodeJS and they seemed to be similar except that they didn't require metadata in the payload.

If the editor needs to be integrated into a different backend, we need to ensure how the first API payload/response must be structured. We can provide detailed documentation for the same.

cc: @karthiknmenon

**Comment by AbhayVAshokan** [comment]
If need be I shall remove the `@rails/activestorage` package completely and add the checksum generation logic to the editor. At the same time the package itself is not bulky since it has a single dependency.

https://github.com/rails/rails/blob/b8fc7898fe885a5243681cf58f67df8bc7920e59/activestorage/package.json#L21-L23

**Review by labeebklatif — APPROVED** [review]
This LGTM.

**Comment by neerajsingh0101** [comment]
I’m late to this but in future I want to be clear that we are building
neeto tools for neeto and not for others. Just because it’s open source we
are not obligated to make it work for non Rails applications. Let’s keep it
simple neeto tools are for neeto. No need to bring any complexity if
someone is using outside of neeto unless it’s of vital interest to us.

**Pull request merged**